### PR TITLE
added sd_watchdog_enabled (#1)

### PR DESCRIPTION
**Pull request by JanZachmann**
* added watchdog_enabled function

* added tests

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -224,13 +225,70 @@ fn fd_cloexec(fd: u32) -> io::Result<()> {
...
+
+    if let Some(t) = s
+        .to_owned()
+        .and_then(|s| s.to_str().and_then(|s| u64::from_str(s).ok()))
+    {
+        if let Some(pid) = p
+            .to_owned()
+            .and_then(|s| s.to_str().and_then(|s| u32::from_str(s).ok()))
+        {
+            if process::id() == pid {
```
Comment (lnicola) [inline src/lib.rs:274]: This doesn't look right. The docs say:

> If the `$WATCHDOG_USEC` environment variable is set, and the `$WATCHDOG_PID` variable is unset or set to the PID of the current process, the service manager expects notifications from this process.

So if `$WATCHDOG_PID` is unset and `$WATCHDOG_USEC` is set, the watchdog is actually enabled, so we should return `true`.

I'd also be inclined to drop the `Result`, or to return `Err` if parsing the two values fails.

File: src/lib.rs [hunk 2]
```diff
@@ -299,4 +357,49 @@ mod tests {
         assert!(env::var_os("LISTEN_PID").is_none());
         assert!(env::var_os("LISTEN_FDS").is_none());
     }
+
+    #[test]
+    fn watchdog_enabled() {
+        // test original logic: https://github.com/systemd/systemd/blob/f3376ee8fa28aab3f7edfad1ddfbcceca5bc841c/src/libsystemd/sd-daemon/sd-daemon.c#L632
```
Comment (lnicola) [inline src/lib.rs:363]: I didn't open this file, so I don't know how close your implementation is to the original wrt. unset `WATCHDOG_PID`, but I'm slightly wary of potential licensing implications here. This crate is licensed as MIT/Apache-2.0, while `libsystemd` is LGPL.

Your implementation of `watchdog_enabled` seems almost trivial, but I'm wary to look, or even link to the `systemd` implementation of it. And if we do add a link to it, it should probably be in the implementation, not the test code.

**Comment by lnicola** [comment]
Sorry, I kinda' forgot about this. I left a couple of remarks, but it's fine if you can't update the PR right now. I can merge it as-is today or tomorrow and fix these up in a later commit.

File: src/lib.rs (hunk 2 above)
Comment (JanZachmann) [reply src/lib.rs:363]: agree, we should better remove the link.

File: src/lib.rs (hunk 1 above)
Comment (JanZachmann) [reply src/lib.rs:274]: You might be right. I didn't look on the spec so much. Instead I tried to implement the same logic provided by the original [code](https://github.com/systemd/systemd/blob/f3376ee8fa28aab3f7edfad1ddfbcceca5bc841c/src/libsystemd/sd-daemon/sd-daemon.c#L632):
```
_public_ int sd_watchdog_enabled(int unset_environment, uint64_t *usec) {
        const char *s, *p = ""; /* p is set to dummy value to do unsetting */
        uint64_t u;
        int r = 0;

        s = getenv("WATCHDOG_USEC");
        if (!s)
                goto finish;

        r = safe_atou64(s, &u);
        if (r < 0)
                goto finish;
        if (u <= 0 || u >= USEC_INFINITY) {
                r = -EINVAL;
                goto finish;
        }

        p = getenv("WATCHDOG_PID");
        if (p) {
                pid_t pid;

                r = parse_pid(p, &pid);
                if (r < 0)
                        goto finish;

                /* Is this for us? */
                if (getpid_cached() != pid) {
                        r = 0;
                        goto finish;
                }
        }

        if (usec)
                *usec = u;

        r = 1;

finish:
        if (unset_environment && s)
                assert_se(unsetenv("WATCHDOG_USEC") == 0);
        if (unset_environment && p)
                assert_se(unsetenv("WATCHDOG_PID") == 0);

        return r;
}
```

**Comment by JanZachmann** [comment]
> Sorry, I kinda' forgot about this. I left a couple of remarks, but it's fine if you can't update the PR right now. I can merge it as-is today or tomorrow and fix these up in a later commit.

I really would like to come back to your offer to merge and fix later on.

**Comment by lnicola** [comment]
Sure, and thanks!

**Comment by JanZachmann** [comment]
> Sure, and thanks!

You're welcome! When can I expect a new tag I can reference in my code?

**Comment by lnicola** [comment]
In the next day or two. I'm somewhat worried about the return type, it should use a custom error enum, but the other methods return `std::io::Result`, which is a bit ugly.

**Comment by JanZachmann** [comment]
> In the next day or two. I'm somewhat worried about the return type, it should use a custom error enum, but the other methods return `std::io::Result`, which is a bit ugly.

Hi. Any news about releasing to crates.io?

**Comment by lnicola** [comment]
Yeah, sorry, I'm this great at keeping deadlines.

I've just made the changes mentioned above and released the new version (`0.4.1`).

**Comment by JanZachmann** [comment]
> Yeah, sorry, I'm this great at keeping deadlines.
> 
> I've just made the changes mentioned above and released the new version (`0.4.1`).

Thx

**Pull request merged**